CLOUDSTACK-9976: Redirect saml2 failed login message to a configurable URL

GabrielBrascher opened the pull request:
When a user fails to authenticate with SAML2, it returns an error page
showing the content of the following image.

![samlloginresponse](https://user-images.githubusercontent.com/5025148/28230557-e1d96b4c-68bd-11e7-93d4-e829de9cf289.png)


To make it more user-friendly and customizable, one could configure a
desirable URL to redirect when such authentication failure happens.


This PR proposes a global settings variable
(saml2.failed.login.redirect.url). If null/empty, the SAML2 authentication
flow does not change from the current; however, if the user configures
a URL then ACS redirects to that URL.

rafaelweingartner commented on plugins/user-authenticators/saml2/src/org/apache/cloudstack/api/command/SAML2LoginAPIAuthenticatorCmd.java:
```diff
@@ -303,9 +308,13 @@ public String authenticate(final String command, final Map<String, Object[]> par
-                            "Your authenticated user is not authorized for SAML Single Sign-On, please contact your administrator",
-                            params, responseType));
+                    String saml2FailedLoginRedirectUrl = configDao.getValue(Config.Saml2FailedLoginRedirectUrl.toString());
+                    if (StringUtils.isEmpty(saml2FailedLoginRedirectUrl)) {
```
@GabrielBrascher what about using isBlank instead of IsEmpty?

GabrielBrascher replied:
Thanks for the hint @rafaelweingartner . I will change it.

rafaelweingartner commented on setup/db/db/schema-4930to41000.sql:
```diff
@@ -254,4 +254,7 @@ CREATE TABLE `cloud`.`firewall_rules_dcidrs`(
\ No newline at end of file
+)ENGINE=InnoDB DEFAULT CHARSET=utf8;
+
+-- CLOUDSTACK-9975: Allow customizing system VM templates for SSVM, VR, and Console Proxy
```
4.10 was already closed, right?
I think you should get the master and add it to 4.10 - 4.11? Or something like that

rafaelweingartner commented on plugins/user-authenticators/saml2/src/org/apache/cloudstack/api/command/SAML2LoginAPIAuthenticatorCmd.java:
```diff
@@ -303,9 +308,13 @@ public String authenticate(final String command, final Map<String, Object[]> par
                 }
 
                 if (userAccount == null || userAccount.getExternalEntity() == null || !_samlAuthManager.isUserAuthorized(userAccount.getId(), issuer.getValue())) {
```
This method is already quite big. What about extracting what you added lines 310-318 to a method?
This enables, docs, test case and a better and clear code to read (in my opinion).

cloudmonger commented:
### ACS CI BVT Run
 **Sumarry:**
 Build Number 973
 Hypervisor xenserver
 NetworkType Advanced
 Passed=107
 Failed=6
 Skipped=12

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/r2si930m8xxzavs/AAAzNrnoF1fC3auFrvsKo_8-a?dl=0

**Failed tests:**
* test_deploy_vm_iso.py

 * test_deploy_vm_from_iso Failing since 31 runs

* test_vm_life_cycle.py

 * test_10_attachAndDetach_iso Failing since 31 runs

* test_routers_network_ops.py

 * test_01_isolate_network_FW_PF_default_routes_egress_true Failing since 64 runs

 * test_02_isolate_network_FW_PF_default_routes_egress_false Failing since 64 runs

 * test_01_RVR_Network_FW_PF_SSH_default_routes_egress_true Failing since 63 runs

 * test_02_RVR_Network_FW_PF_SSH_default_routes_egress_false Failing since 63 runs


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_01_verify_libvirt
test_02_verify_libvirt_after_restart
test_03_verify_libvirt_attach_disk
test_04_verify_guest_lspci
test_05_change_vm_ostype_restart
test_06_verify_guest_lspci_again
test_static_role_account_acls
test_11_ss_nfs_version_on_ssvm
test_nested_virtualization_vmware
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_vm_snapshots.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_loadbalance.py
test_routers.py
test_reset_vm_on_reboot.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_non_contigiousvlan.py
test_login.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_metrics_api.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_volumes.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_disk_offerings.py

GabrielBrascher commented:
Thanks for the feedback @rafaelweingartner.
I updated the SQL from file schema-4930 to 41000.sql to schema-41000to41100.sql, extracted the code from SAML2LoginAPIAuthenticatorCmd.java, added documentation, and unit tests.

cloudmonger commented:
### ACS CI BVT Run
 **Sumarry:**
 Build Number 991
 Hypervisor xenserver
 NetworkType Advanced
 Passed=107
 Failed=6
 Skipped=12

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/r2si930m8xxzavs/AAAzNrnoF1fC3auFrvsKo_8-a?dl=0

**Failed tests:**
* test_deploy_vm_iso.py

 * test_deploy_vm_from_iso Failing since 49 runs

* test_vm_life_cycle.py

 * test_10_attachAndDetach_iso Failing since 49 runs

* test_routers_network_ops.py

 * test_01_isolate_network_FW_PF_default_routes_egress_true Failing since 82 runs

 * test_02_isolate_network_FW_PF_default_routes_egress_false Failing since 82 runs

 * test_01_RVR_Network_FW_PF_SSH_default_routes_egress_true Failing since 80 runs

 * test_02_RVR_Network_FW_PF_SSH_default_routes_egress_false Failing since 80 runs


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_01_verify_libvirt
test_02_verify_libvirt_after_restart
test_03_verify_libvirt_attach_disk
test_04_verify_guest_lspci
test_05_change_vm_ostype_restart
test_06_verify_guest_lspci_again
test_static_role_account_acls
test_11_ss_nfs_version_on_ssvm
test_nested_virtualization_vmware
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_vm_snapshots.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_loadbalance.py
test_routers.py
test_reset_vm_on_reboot.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_non_contigiousvlan.py
test_login.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_metrics_api.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_volumes.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_disk_offerings.py

cloudmonger commented:
### ACS CI BVT Run
 **Sumarry:**
 Build Number 992
 Hypervisor xenserver
 NetworkType Advanced
 Passed=103
 Failed=10
 Skipped=12

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/r2si930m8xxzavs/AAAzNrnoF1fC3auFrvsKo_8-a?dl=0

**Failed tests:**
* test_loadbalance.py

 * test_01_create_lb_rule_src_nat Failed

 * test_02_create_lb_rule_non_nat Failed

 * test_assign_and_removal_lb Failed

* test_deploy_vm_iso.py

 * test_deploy_vm_from_iso Failing since 50 runs

* test_volumes.py

 * test_06_download_detached_volume Failed

* test_vm_life_cycle.py

 * test_10_attachAndDetach_iso Failing since 50 runs

* test_routers_network_ops.py

 * test_01_isolate_network_FW_PF_default_routes_egress_true Failing since 83 runs

 * test_02_isolate_network_FW_PF_default_routes_egress_false Failing since 83 runs

 * test_01_RVR_Network_FW_PF_SSH_default_routes_egress_true Failing since 81 runs

 * test_02_RVR_Network_FW_PF_SSH_default_routes_egress_false Failing since 81 runs


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_01_verify_libvirt
test_02_verify_libvirt_after_restart
test_03_verify_libvirt_attach_disk
test_04_verify_guest_lspci
test_05_change_vm_ostype_restart
test_06_verify_guest_lspci_again
test_static_role_account_acls
test_11_ss_nfs_version_on_ssvm
test_nested_virtualization_vmware
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_vm_snapshots.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_routers.py
test_reset_vm_on_reboot.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_non_contigiousvlan.py
test_login.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_metrics_api.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_disk_offerings.py

rhtyd commented on server/src/com/cloud/configuration/Config.java:
```diff
@@ -664,6 +664,14 @@
             "The mount point on the Management Server for Secondary Storage.",
             null),
+    Saml2FailedLoginRedirectUrl(
+            "Advanced",
```
@GabrielBrascher please move this to a ConfigKey, existing deployments may not have this config created.

rhtyd commented on setup/db/db/schema-41000to41100.sql:
```diff
@@ -18,3 +18,6 @@
 -- Schema upgrade from 4.10.0.0 to 4.11.0.0
 --;
+
+-- CLOUDSTACK-9976: Redirect saml2 failed login message to a configurable URL
```
We can remove this sql change, if we implement a ConfigKey instead.

GabrielBrascher replied:
Thanks for the hint @rhtyd !

rhtyd commented:
@GabrielBrascher please fix the conflicts and address outstanding issues

rhtyd commented:
@GabrielBrascher please address outstanding comments, and fix the conflicts, thanks.

GabrielBrascher commented:
Sorry for the delay @rhtyd. I will work on that when possible.

GabrielBrascher commented:
Refactored using ConfigKey, as you suggested @rhtyd. Thanks for all the feedback!

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✖debian. JID-1116

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✖debian. JID-1117

GabrielBrascher commented:
I will check these failures.

rhtyd commented:
@GabrielBrascher ping, update on this PR? (looks like build failure)

GabrielBrascher commented:
@rhtyd I will check on that.

rhtyd commented:
@GabrielBrascher ping?

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✔centos7 ✖debian. JID-1410

GabrielBrascher commented:
@rhtyd I am checking them, thanks.

rhtyd commented:
Can you verify/fix the failures @GabrielBrascher

rhtyd commented:
Removing this from 4.11 due to lack of engagement from the PR author.

GabrielBrascher commented:
@rhtyd rebased against latest master. Sorry for taking too long.

GabrielBrascher commented:
@rhtyd all checks have passed.

rafaelweingartner commented:
@blueorangutan package

blueorangutan commented:
@rafaelweingartner a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-2320

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

rafaelweingartner commented on plugins/user-authenticators/saml2/src/main/java/org/apache/cloudstack/api/command/SAML2LoginAPIAuthenticatorCmd.java:
```diff
@@ -329,6 +333,24 @@ public String authenticate(final String command, final Map<String, Object[]> par
+    protected void whenFailToAuthenticateThrowExceptionOrRedirectToUrl(final Map<String, Object[]> params, final String responseType, final HttpServletResponse resp, Issuer issuer,
+            UserAccount userAccount) throws IOException {
+        if (userAccount == null || userAccount.getExternalEntity() == null || !_samlAuthManager.isUserAuthorized(userAccount.getId(), issuer.getValue())) {
+            String saml2FailedLoginRedirectUrl = (String)getConfigKeys()[0].value();
```
What about using `Saml2FailedLoginRedirectUrl.value()` instead of `getConfigKeys()[0].value()`?

DaanHoogland commented:
@rhtyd are you alright with the contents of this ?
